crates/durability/src/imp/local.rs

```diff
@@ -0,0 +1,300 @@
+#[derive(Clone, Copy, Debug)]
+pub struct Options {
+    /// Periodically flush and sync the log this often.
+    ///
+    /// Default: 500ms
+    pub sync_interval: Duration,
+    /// [`Commitlog`] configuration.
+    pub commitlog: spacetimedb_commitlog::Options,
+}
+
+impl Default for Options {
+    fn default() -> Self {
+        Self {
+            sync_interval: Duration::from_millis(500),
+            commitlog: Default::default(),
+        }
+    }
+}
+
+/// [`Durability`] implementation backed by a [`Commitlog`] on local storage.
+///
+/// The commitlog is constrained to store the canonical [`Txdata`] payload,
+/// where the generic parameter `T` is the type of the row data stored in
+/// the mutations section.
+///
+/// `T` is left generic in order to allow bypassing the `ProductValue`
+/// intermediate representation in the future.
+///
+/// Note, however, that instantiating `T` to a different type may require to
+/// change the log format version!
+pub struct Local<T> {
+    /// The [`Commitlog`] this [`Durability`] and [`History`] impl wraps.
+    clog: Arc<Commitlog<Txdata<T>>>,
+    /// The durable transaction offset, as reported by the background
+    /// [`FlushAndSyncTask`].
+    ///
+    /// A negative number indicates that we haven't flushed yet, or that the
+    /// number overflowed. In either case, appending new transactions shall panic.
+    ///
+    /// The offset will be used by the datastore to squash durable transactions
+    /// into the committed state, thereby making them visible to durable-only
+    /// readers.
+    ///
+    /// We don't want to hang on to those transactions longer than needed, so
+    /// acquire / release or stronger should be used to prevent stale reads.
+    durable_offset: Arc<AtomicI64>,
+    /// Backlog of transactions to be written to disk by the background
+    /// [`PersisterTask`].
+    ///
+    /// Note that this is unbounded!
+    queue: mpsc::UnboundedSender<Txdata<T>>,
+    /// How many transactions are sitting in the `queue`.
+    ///
+    /// This is mainly for observability purposes, and can thus be updated with
+    /// relaxed memory ordering.
+    queue_depth: Arc<AtomicU64>,
+    /// Handle to the [`PersisterTask`], allowing to drain the `queue` when
+    /// explicitly dropped via [`Self::close`].
+    persister_task: JoinHandle<()>,
+}
+
+impl<T: Encode + Send + Sync + 'static> Local<T> {
+    /// Create a [`Local`] instance at the `root` directory.
+    ///
+    /// The `root` directory must already exist.
+    ///
+    /// Background tasks are spawned onto the provided tokio runtime.
+    pub fn open(root: impl Into<PathBuf>, rt: tokio::runtime::Handle, opts: Options) -> io::Result<Self> {
+        info!("open local durability");
+
+        let clog = Arc::new(Commitlog::open(root, opts.commitlog)?);
+        let (queue, rx) = mpsc::unbounded_channel();
+        let queue_depth = Arc::new(AtomicU64::new(0));
+        let offset = Arc::new(AtomicI64::new(-1));
```

Comment:
Shouldn't we be initializing this based on the clog's offset on disk after opening rather than `-1`? Am I misunderstanding?

Reply:
Yeah can do.

I'll note, however, that it is basically non-sensical to call `durable_tx_offset` before anything was written by the owner of the `Durability`: we already know the offset from replaying the `History` (which, n.b., does not need to be the same source as the `Durability`).

Things would look a little different if `append_tx` would take the transaction offset, but then the question would be: what happens if one tries to `append_tx` with an offset that is _not_ `n + 1`?

Unfortunately, we have to keep track of the offset on both sides -- datastore and log.